### PR TITLE
Multi-Label Classification with Scikit-MultiLearn

### LINKED ISSUE #3518: Multi-Label Classification with Scikit-MultiLearn
### Proposed title of article
Multi-Label Classification with Scikit-MultiLearn

### Introduction paragraph (2-3 paragraphs):
Multi-label classification methods allow us to classify data sets with more than 1 target variable, a prediction can belong to various labels. There are various methods that should be used depending on the dataset on hand.
Methods for solving Multi-label Classification Problems
1. Problem Transformation
2. Adapted Algorithm
3. Ensemble approaches

1. Problem Transformation
It refers to transforming the multi-label problem into a single-label problem(s) by using
- Binary Relevance: treats each label as a separate single class classification
- Classifier Chains: In this, the first classifier is trained just on the input data and then each next classifier is trained on the input space and all the previous classifiers in the chain.
- Label Powerset: we transform the problem into a multi-class problem with one multi-class classifier is trained on all unique label combinations found in the training data.
2. Adapted Algorithm
adapting the algorithm to directly perform multi-label classification, rather than transforming the problem into different subsets of problems.
3. Ensemble methods.
Incorporate the merits of two previous approaches.

In this tutorial, we shall use all three methods in building our multi-label model using Scikit-Multilearn.
Scikit-multilearn is a BSD-licensed library for multi-label classification that is built on top of the well-known scikit-learn ecosystem.
### Key takeaways: 
1. Difference between Multi-class classification and Multi-label classification.
2. Installing and exploring Scikit-Multilearn.
3. Using all the three approaches in building and training our model.
4. Building a text classification model.
5. Finding the optimal method of the three.

### References: 
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by charles721**
This fixes #3518 

### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #issue-number`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
plagiarism check done

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 1]
```diff
@@ -0,0 +1,654 @@
+Multi-label classification allows us to classify data sets with more than one target variable. In multi-label classification, we have several labels that are the outputs for a given prediction. When making predictions, a given input may belong to more than one label.
+
+For example, when predicting a given movie category, it may belong to horror, romance, adventure, action, or simultaneously. In this example, we have multi-labels that can be assigned to a given movie. In multi-class classification, an input belongs to only a single label. For example, when predicting if a given image belongs to a cat or a dog, the output can be either a cat or dog but not both at the same time.
+
+In this tutorial, we will be dealing with multi-label text classification, and we will build a model that classifies a given text input into different categories. Our text input can belong to multiple categories or labels at the same time.
+
+We will use scikit-multi learn in building our model. Scikit-multi learn is a python library built on top of scikit-learn and is best suited for multi-label classification.
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:8]:
```suggestion
We will use scikit-multilearn in building our model. Scikit-multilearn is a python library built on top of scikit-learn and is best suited for multi-label classification.
```

Comment (Bayler) [reply content/articles/multi-label-classification-with-scikit-multilearn/index.md:8]: Also add an external link to Multi-scikitlearn documentation.

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 2]
```diff
@@ -0,0 +1,654 @@
...
+We check the structure to be able to see the available columns in our dataset.
+
+```python
+df.head()
+```
+
+The output is as shown.
+
+![Dataset structure](/engineering-education/multi-label-classification-with-scikit-multilearn/dataset-structure.jpg)
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:157]: What do we decode from these results? Explain to the reader what these outputs mean

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 3]
```diff
@@ -0,0 +1,654 @@
...
+1. A good understanding of [Python](/engineering-education/python-projects-for-beginners/).
+2. A good understanding of [machine learning modeling](/engineering-education/house-price-prediction/).
+3. Download the dataset [here](https://drive.google.com/file/d/1_scJyWek8DJUZ14_wZ8GXZslpwT8ubzt/view?usp=sharing).
+4. Have some knowledge of [Pandas](https://numpy.org/) and [Numpy.](https://numpy.org/)
+   > NOTE: To follow along easily, use [Google Colab](https://research.google.com/) to build your model.
+
+### Introduction
+
+We have different methods and techniques used for multi-label classification. These methods are as follows.
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:50]: Please write a brief intro, maybe a paragraph or two, before diving into the different methods needed.

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 4]
```diff
@@ -0,0 +1,654 @@
...
+When using this technique, label correlation may be lost; hence we introduce a second technique.
+
+#### Classifier chains
+
+In this technique, we have multiple classifiers connected in a chain. The first classifier is build using the input data. The following classifiers are trained using the combined inputs and the previous classifiers in a given chain.
+
+This is a sequential process where an output of one classifier is used as the input of the next classifier in the chain, as shown in the image below.
+
+![Classifier chain sequence](/engineering-education/multi-label-classification-with-scikit-multilearn/chain-sequence.png)
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:85]: Are these images yours? If not, please add their source information

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 5]
```diff
@@ -0,0 +1,654 @@
...
+Text preprocessing involves removing stop words and removal of noisy data. Stop words are a list of all words that are common in a given language.
+
+Stop words are removed since they do not have a high classification power during predictive analysis. They tend to bring bias when building the classifier.
+
+We remove noisy data that affect our model during training and bring errors in our model.
+
+To perform text preprocessing, we need to install the neat text package. Neat text is a Python package that is used for textual data cleaning and text preprocessing.
+
+#### Neat text installating
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:262]: Make sure to always proofread your work. This should be 'installation'

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 6]
```diff
@@ -0,0 +1,654 @@
...
+Let's install using the following command.
+
+```python
+!pip install neattext
+```
+
+Let's load this package.
+
+#### Loading neat text
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:272]: I believe neattext is a single word, right? Correct in all areas where you've added space

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 7]
```diff
@@ -0,0 +1,654 @@
...
+### Feature engineering
+
+Feature engineering involves extracting features and properties from our data. Features are the independent units that are used for predictive analysis to influence the output. We shall use the `TfidfVectorizer()` package to conduct feature extraction, which we imported earlier.
+
+#### TfidfVectorizer initilization
+
+```python
+tfidf = TfidfVectorizer()
+```
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:376]: Has this code extracted the features? This section is not clear

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 8]
```diff
@@ -0,0 +1,654 @@
...
+```python
+y = df[['mysql', 'python', 'php']]
+```
+
+### Dataset split
+
+We split the dataset into train set and test set. We will use the train set to build our model while the text set is used to gauge the model performance.
+
+We use the `train_test_split` to split our dataset into two. 70% of the split dataset is used as the train set, and 30% is used as the test set.
+
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:414]:
```suggestion
We use the `train_test_split` to split our dataset into two. 70% of the dataset is used as the train set, and 30% is used as the test set.
```

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 9]
```diff
@@ -0,0 +1,654 @@
...
+As shown, it uses the `MultinomialNB()` to handle the converted multi-class problem.
+
+#### Testing our model
+
+We test our model using the test data. We see if the model can make a prediction using the test data.
+
+```pyhthon
+br_prediction = binary_rel_clf.predict(X_test)
+```
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:461]:
```pyhthon

Please proofread your work

File: content/articles/multi-label-classification-with-scikit-multilearn/index.md [hunk 10]
```diff
@@ -0,0 +1,654 @@
...
+
+```python
+hamming_loss(y_test,br_prediction)
+```
+
+The output is as shown.
+
+```bash
+0.06060606060606061
+```
```
Comment (Bayler) [inline content/articles/multi-label-classification-with-scikit-multilearn/index.md:516]: What does this result mean? Is the hamming loss low? Explain the result for the benefit of the reader

**Comment by Bayler** [comment]
Hi @charles721 This was a nice read. Being your first article, I am impressed by how well you're explaining concepts. Just some minor revisions then we can continue for final review.

**Comment by charles721** [comment]
Hello @Bayler, I have made the requested changes
Thanks

**Comment by Bayler** [comment]
> Hello @Bayler, I have made the requested changes
> Thanks

Noted. Let me have a look at it

**Comment by Bayler** [comment]
@charles721 Please check your tutorial once again and confirm to me that all the remaining images that aren't yours have been added sources. Once confirmed, let me know so that we can move for final review.

**Comment by charles721** [comment]
Hello @Bayler , the remaining images without image source are mine, which I created to explain the concept of binary relevance, classifier chains and label powerset. Others are screenshots from outputs.
I only used one external image which I added an image source.
Thanks.

**Comment by Bayler** [comment]
@charles721 This is a great read. Being your first PR, I'm impressed by your level of writing. Keep up the good work and see you in the next one.

**Comment by charles721** [comment]
Thanks for the review @Bayler

**Comment by LinusMuema** [comment]
Hi @charles721 
Thanks for this amazing article and for joining the section community. We look forward to getting more content from you.

cc: @Bayler check my commits for more changes. Also, the conclusion was quite lengthy while it's supposed to be a short summary of the article. Also, check the use of semi-colons and general punctuation. Otherwise, awesome job!
